Issue2099

lklivingstone opened the pull request:
**Help needed in the following**
- [ ] Contributions to qutip should follow the [pep8 style](https://www.python.org/dev/peps/pep-0008/).

>I did the `pycodestyle` on my code, but it is highlighting errors that were there in the original code. It returned lines that have >linespace errors and more, should I edit them too?
>Output:

> ```
> D:\qutip\qutip\core\cy>pycodestyle --first qobjevo.pyx
> qobjevo.pyx:1:1: E265 block comment should start with '# '
> qobjevo.pyx:2:80: E501 line too long (83 > 79 characters)
> qobjevo.pyx:20:37: E211 whitespace before '('
> qobjevo.pyx:46:56: W605 invalid escape sequence '\s'
> qobjevo.pyx:190:30: E225 missing whitespace around operator
> qobjevo.pyx:423:5: E303 too many blank lines (2)
> qobjevo.pyx:426:5: E301 expected 1 blank line, found 0
> qobjevo.pyx:451:34: E127 continuation line over-indented for visual indent
> qobjevo.pyx:572:21: E128 continuation line under-indented for visual indent
> qobjevo.pyx:987:39: E231 missing whitespace after ','
> qobjevo.pyx:989:20: E124 closing bracket does not match visual indentation
> ```

- [ ] Please add tests to cover your changes if applicable.

> The issue was to add `__repr__` function to `QobjEvo` class. Do I need to add a test? If so, can someone help me with it?

- [ ] If the behavior of the code has changed or new feature has been added, please also update the documentation in the `doc` folder, and the [notebook](https://github.com/qutip/qutip-tutorials). Feel free to ask if you are not sure.

> I am not sure in what bracket this issue falls. The issue was labeled as `ENH`.

- [ ] Include the changelog in a file named: `doc/changes/<PR number>.<type>` 'type' can be one of the following: feature, bugfix, doc, removal, misc, or deprecation (see [here](http://qutip.org/docs/latest/development/contributing.html#changelog-generation) for more information).

> Related to the previous point.

**Description**
Added `__repr__` to `QobjEvo`

It is showing the following attributes: `dims`, `shape`, `type`, `superop` (if a super operator is present), `isconstant` and `num_elements`.

Output:
```
>>> import qutip
>>> qutip.QobjEvo([qutip.qeye(2), lambda t: t])
<QobjEvo dims = [[2], [2]], shape= (2, 2), type = oper, superrep = None, isconstant = False, num_elements = 1>
```



**Related issues or PRs**
Fix #2099

AGaliciaMartinez commented:
> I did the pycodestyle on my code, but it is highlighting errors that were there in the original code. It returned lines that have >linespace errors and more, should I edit them too?

I am going to leave this up to you. It is not required as long as the lines you modified follow PEP8. Althoug, fixing the output of pycodestyle in the same file would be very much welcome :smile:.

> The issue was to add __repr__ function to QobjEvo class. Do I need to add a test? If so, can someone help me with it?

We will indeed need a test for repr. It does not need to be very complicated. In this case, we would create a few different QobjEvo that cover most of the cases and we would check that `qobjevo.__rerp__()` returns the expected string. This test should go in `qutip/tests/core/test_qobjevo.py`. You can take a look there to see example of how the tests are writtent. If this is your first time writing tests you may want to take a look at [our documentation](https://qutip.readthedocs.io/en/master/development/contributing.html#testing) and also at pytests documentation.

> If the behavior of the code has changed or new feature has been added, please also update the documentation in the doc folder, and the [notebook](https://github.com/qutip/qutip-tutorials). Feel free to ask if you are not sure.

This will not be necessary

> Include the changelog in a file named: doc/changes/<PR number>.<type> 'type' can be one of the following: feature, bugfix, doc, removal, misc, or deprecation (see [here](http://qutip.org/docs/latest/development/contributing.html#changelog-generation) for more information).

`feature` should work here.

AGaliciaMartinez commented on qutip/core/cy/qobjevo.pyx:
```diff
@@ -231,6 +231,10 @@ cdef class QobjEvo:
 
+    def __repr__(self):
+        cls = self.__class__.__name__
+        return f'<{cls} dims = {self.dims}, shape= {self.shape}, type = {self.type}, superrep = {self.superrep}, isconstant = {self.isconstant}, num_elements = {self.num_elements}>'
```
```suggestion
        return f'{cls}: dims={self.dims}, shape={self.shape}, type={self.type}, superrep={self.superrep}, isconstant={self.isconstant}, num_elements={self.num_elements}>'
```
I changed slightly the format to follow the same as for Qobj.

AGaliciaMartinez reviewed (changes requested):
thanks for your contribution! I added one suggestion but we would need some tests to be added as explained above before this gets merged. Feel free to ask any question!

lklivingstone commented:
I am working on it right now and will commit to the repo with all the changes.

lklivingstone commented:
@AGaliciaMartinez Hello, I have made the required changes in the __repr__() of QobjEvo. 

Output of the new code:
```
>>> import qutip
>>> qutip.QobjEvo([qutip.qeye(2), lambda t: t])
<QobjEvo: dims=[[2], [2]], shape=(2, 2), type=oper, superrep=None, isconstant=False, num_elements=1>
```
_In your suggestion, there wasnt a "<" before QobjEvo, so I added that._

I have also edited the page according to the `pycodestyle`.

Kindly check the test too. This is my first test, so do suggest changes, if required. I could not think of different cases, so I write 4 cases, and checked them according to the expected result.

AGaliciaMartinez commented:
Also, regarding the `<`, I purposefully did not include to make the print similar to Qobj __repr__ 😉.

lklivingstone commented:
@AGaliciaMartinez Thank you!

lklivingstone commented:
@AGaliciaMartinez Thank you!
I have removed both `<` and `>` from the `repr()`.

Now it looks like this:
```
>>> import qutip
>>> qutip.QobjEvo([qutip.qeye(2), lambda t: t])
QobjEvo: dims=[[2], [2]], shape=(2, 2), type=oper, superrep=None, isconstant=False, num_elements=1
```

I have added many tests according to your suggestions. Now there are 7 cases to check.

Ericgig commented on qutip/core/cy/qobjevo.pyx:
```diff
@@ -1,5 +1,6 @@
-#cython: language_level=3
-#cython: boundscheck=False, wraparound=False, initializedcheck=False, cdvision=True
+# cython: language_level=3
+# cython: boundscheck=False, wraparound=False,
+#         initializedcheck=False, cdvision=True
```
These are cython options, not comments.
Splitting it like this change the behavior!

Ericgig commented on qutip/core/cy/qobjevo.pyx:
```diff
@@ -17,7 +18,7 @@ from qutip.settings import settings
 from qutip.core.data cimport Dense, Data, dense
 from qutip.core.data.expect cimport *
-from qutip.core.data.reshape cimport (column_stack_dense, column_unstack_dense)
+from qutip.core.data.reshape cimport(column_stack_dense, column_unstack_dense)
```
`cimport` is not a function, there should be a space here.

Ericgig commented on qutip/tests/core/test_qobjevo.py:
```diff
@@ -131,6 +132,37 @@ def test_call(pseudo_qevo, coeff_type):
+    expected_repr_1= 'QobjEvo: dims=[[3], [3]], shape=(3, 3), type=oper, superrep=None, isconstant=False, num_elements=1'
+    assert case_1 == expected_repr_1
+
+    case_2= repr(QobjEvo(qeye(2), lambda t: t))
```
This case is strange since it's equivalent to `QobjEvo(qeye(2), args=lambda t: t)` and the `args` are not used...
It simply look as a badly formatted `QobjEvo`.

Ericgig reviewed (changes requested):
Be careful when applying `pycodestyle` to cython code. cython has new keywords that it does not knows.

lklivingstone commented:
Changes to the test:
```
    case_1= repr(QobjEvo([qeye(3), lambda t: t]))
    expected_repr_1= 'QobjEvo: dims=[[3], [3]], shape=(3, 3), type=oper, superrep=None, isconstant=False, num_elements=1'
    assert case_1 == expected_repr_1

    case_2= repr(QobjEvo(qeye(2)))
    expected_repr_2= 'QobjEvo: dims=[[2], [2]], shape=(2, 2), type=oper, superrep=None, isconstant=True, num_elements=1'
    assert case_2 == expected_repr_2
    
    case_3= repr(QobjEvo(basis(5, 2)))
    expected_repr_3= 'QobjEvo: dims=[[5], [1]], shape=(5, 1), type=ket, superrep=None, isconstant=True, num_elements=1'
    assert case_3 == expected_repr_3

    X = sigmax()
    S = spre(X) * spost(X.dag())
    case_4= repr(QobjEvo(to_choi(S)))
    expected_repr_4= 'QobjEvo: dims=[[[2], [2]], [[2], [2]]], shape=(4, 4), type=super, superrep=choi, isconstant=True, num_elements=1'
    assert case_4 == expected_repr_4

    case_5= repr(QobjEvo([[qeye(4), lambda t: t], [qeye(4), lambda t: t]], compress=False))
    expected_repr_5= 'QobjEvo: dims=[[4], [4]], shape=(4, 4), type=oper, superrep=None, isconstant=False, num_elements=2'
    assert case_5 == expected_repr_5
```

Changes to the `cython` code have also been undone.

@Ericgig @AGaliciaMartinez Kindly let me know if there are any more changes to be done.

lklivingstone commented:
@Ericgig I resolved the conversation. Do I need to do anything else?

Ericgig commented:
It's not the conversation but the conflict in the `qobjevo.pyx` file that need to be resolved.
The PR cannot be merged while there is a conflict. Most of the tests can't be ran.

Ericgig commented on qutip/core/cy/qobjevo.pyx:
```diff
@@ -241,6 +241,10 @@ cdef class QobjEvo:
 
+    def __repr__(self):
+        cls = self.__class__.__name__
+        return f'{cls}: dims={self.dims}, shape={self.shape}, type={self.type}, superrep={self.superrep}, isconstant={self.isconstant}, num_elements={self.num_elements}'
```
This line is quite wide, can you split it?

Ericgig commented on qutip/tests/core/test_qobjevo.py:
```diff
@@ -131,6 +132,33 @@ def test_call(pseudo_qevo, coeff_type):
     assert not qevo.issuper
     _assert_qobjevo_equivalent(pseudo_qevo, qevo)
 
+# Test the QobjEvo.__repr__()
```
The pep8 of the test is not great.

Ericgig reviewed (commented):
A few lines are too long, but the code is good.

lklivingstone commented:
I have made corrections to the __lines of code which I added__ in the file `test_qobjevo.py`. 
I have also changed the format of `__repr__` with proper spacing.

__repr__:
```
>>> import qutip
>>> qutip.QobjEvo([qutip.qeye(2), lambda t: t])
QobjEvo: dims = [[2], [2]], shape = (2, 2), type = oper, superrep = None, isconstant = False, num_elements = 1
>>>
```
